### DISCUSSION TITLE
How to remove 3D object/model

### DISCUSSION THREAD

**Discussion by Mato42sk** [sceneview/sceneform-android#276] [Q&A] [answered]
How to remove an object (model.glb) , which is added from library and you want to remove it with a click on that object or after button click and after remove u want to insert new for example second object from library.
In Kotlin.

**Comment by grassydragon** [comment, marked as answer]
- Add the first object when it is selected from the library
- Remove the first object (everything together with the `AnchorNode`) as described here: https://sceneview.github.io/sceneform-android/remove_node
- Add the second object from the library the same way as you do for the first one

**Comment by Mato42sk** [comment]
Yes, but I have a problem that I don't know where I get this AnchorNode from. Look at this. This is function where I am adding objects. I don't know how to access this AnchorNode in code and create removeObject function. Can u help me with it?


 ``` Kotlin
private fun onTapPlane(hitResult: HitResult, plane: Plane, motionEvent: MotionEvent) {
    if (model == null) {
        Toast.makeText(context, "Select a model", Toast.LENGTH_SHORT).show()
        return
    }

    // Create the Anchor.
    scene.addChild(AnchorNode(hitResult.createAnchor()).apply {
        // Create the transformable model and add it to the anchor.
        addChild(TransformableNode(arFragment.transformationSystem).apply {
            renderable = model
            renderableInstance.animate(true).start()
            // Add the View
            addChild(Node().apply {
                // Define the relative position
                localPosition = Vector3(0.0f, 1f, 0.0f)
                localScale = Vector3(0.7f, 0.7f, 0.7f)
            })
        })
    })
}
```

**Reply by grassydragon** [reply]
You can just add a property to your class like `private var anchorNode: AnchorNode? = null` and when store a reference to the added `AnchorNode`:
``` Kotlin
anchorNode = AnchorNode(hitResult.createAnchor()).apply {
    // Create the transformable model and add it to the anchor.
    addChild(TransformableNode(arFragment.transformationSystem).apply {
        renderable = model
        renderableInstance.animate(true).start()
        // Add the View
        addChild(Node().apply {
            // Define the relative position
            localPosition = Vector3(0.0f, 1f, 0.0f)
            localScale = Vector3(0.7f, 0.7f, 0.7f)
        })
    })
}

scene.addChild(anchorNode)
```

**Reply by Mato42sk** [reply]
thanks a lot, but I still have a problem... I can't launch it. Firstly I had this function

``` Kotlin
private fun remove{
    scene.removeChild(anchorNode)
    anchorNode.anchor = null
    anchorNode.parent = null
    anchorNode.renderable = null
}
```
but this shows errors on dots

now I have this, but this shows error too

``` Kotlin
remove_btn.setOnClickListener {
    if (anchorNode != null)
        anchorNode.anchor = null  // here
}
```

I don't know how to fix it...  and create and call a function which will finally works

**Reply by grassydragon** [reply]
The are two solutions. The first one is to make the property `lateinit` like `private lateinit var anchorNode: AnchorNode` but you should be sure that this property is initialized when you use it in your code. In this case your `remove` method can remain unchanged. The second one is to assign the value of `anchorNode` to a local variable so that it can be checked for `null` and used afterwards (this is a Kotlin feature) or use `apply` like:
``` Kotlin
anchorNode?.apply {
    anchor = null
    parent = null
    renderable = null
}

**Reply by Mato42sk** [reply]
ooh thanks a lot for help. I don't want to annoy, but can I ask one more question? Because now works it finally, but it always only last added item, how can I find out which object is selected and then delete the selected object. I tried something, but didn't work and in documetation didn't found

**Reply by grassydragon** [reply]
You can get it from `TransformationSystem` but don't forget that there is a parent `AnchorNode`:
https://github.com/SceneView/sceneform-android/blob/65d6e6ecccabd56fb21c64464793f750d60ccaf7/ux/src/main/java/com/google/ar/sceneform/ux/BaseTransformableNode.java#L53

**Reply by Mato42sk** [reply]
Ommg I appreciate it 💯
